src/Kestrel.Core/Internal/ConnectionHandler.cs

```diff
@@ -34,89 +28,45 @@ public void OnConnection(IFeatureCollection features)
 
             // Set the transport and connection id
-            connectionContext.ConnectionId = connectionId;
-            transportFeature.Connection = new PipeConnection(inputPipe.Reader, outputPipe.Writer);
-            var applicationConnection = new PipeConnection(outputPipe.Reader, inputPipe.Writer);
-
-            if (!_serviceContext.ConnectionManager.NormalConnectionCount.TryLockOne())
-            {
-                var goAway = new RejectionConnection(inputPipe, outputPipe, connectionId, _serviceContext)
-                {
-                    Connection = applicationConnection
-                };
-
-                connectionContext.Features.Set<IConnectionApplicationFeature>(goAway);
-
-                goAway.Reject();
-                return;
-            }
-
-            var frameConnectionContext = new FrameConnectionContext
-            {
-                ConnectionId = connectionId,
-                FrameConnectionId = frameConnectionId,
-                ServiceContext = _serviceContext,
-                PipeFactory = connectionContext.PipeFactory,
-                ConnectionAdapters = _listenOptions.ConnectionAdapters,
-                Input = inputPipe,
-                Output = outputPipe
-            };
-
-            var connectionFeature = connectionContext.Features.Get<IHttpConnectionFeature>();
-
-            if (connectionFeature != null)
-            {
-                if (connectionFeature.LocalIpAddress != null)
-                {
-                    frameConnectionContext.LocalEndPoint = new IPEndPoint(connectionFeature.LocalIpAddress, connectionFeature.LocalPort);
-                }
-
-                if (connectionFeature.RemoteIpAddress != null)
-                {
-                    frameConnectionContext.RemoteEndPoint = new IPEndPoint(connectionFeature.RemoteIpAddress, connectionFeature.RemotePort);
-                }
-            }
-
-            var connection = new FrameConnection(frameConnectionContext)
-            {
-                Connection = applicationConnection
-            };
+            connectionContext.ConnectionId = CorrelationIdGenerator.GetNextId();
+            connectionContext.Transport = pair.Transport;
 
-            connectionContext.Features.Set<IConnectionApplicationFeature>(connection);
+            // This *must* be set before returning from OnConnection
+            transportFeature.Application = pair.Application;
 
-            // Since data cannot be added to the inputPipe by the transport until OnConnection returns,
-            // Frame.ProcessRequestsAsync is guaranteed to unblock the transport thread before calling
-            // application code.
-            connection.StartRequestProcessing(_application);
+            // REVIEW: This task should be tracked by the server for graceful shutdown
+            // Today it's handled specifically for http but not for aribitrary middleware
```

Comment:
Can we at least log a critical error if something goes wrong? I know that shouldn't/won't hapeen today, but this would also help avoid potential unobserved exceptions in the future.